File: ATAK5.6/app/src/main/java/com/atakmap/android/featurelink/arcgis/AutoSymbology.java
package com.atakmap.android.featurelink.arcgis;

import android.graphics.Color;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * ArcGIS renderer -> shape/marker style extractor — sibling to {@link AutoIconset}, which only
 * handles esriPMS (picture-marker) symbols. This extracts esriSMS (point marker shape/color),
 * esriSLS (line stroke), and esriSFS (polygon fill + outline) symbol JSON into plain style
 * structs, keyed by renderer field value the same way AutoIconset.Result.pathByValue is — no
 * network I/O, no install side effects, pure JSON parsing.
 */
public final class AutoSymbology {

    private static final String TAG = "AutoSymbology";

    public static final class StrokeStyle {
        public final int color;
        public final float widthPx;
        public final String dash; // "solid" | "dash" | "dot"

        StrokeStyle(int color, float widthPx, String dash) {
            this.color = color;
            this.widthPx = widthPx;
            this.dash = dash;
        }
    }

    public static final class FillStyle {
        public final int color;
        public final String style; // "solid" | "none"
        /** The fill symbol's own outline (esriSFS.outline is itself an esriSLS) — may be null. */
        public final StrokeStyle outline;

        FillStyle(int color, String style, StrokeStyle outline) {
            this.color = color;
            this.style = style;
            this.outline = outline;
        }
    }

    public static final class MarkerStyle {
        public final int color;
        public final String shape; // "circle" | "square" | "diamond" | "triangle" | "cross" | "x"
        public final float sizePx;

        MarkerStyle(int color, String shape, float sizePx) {
            this.color = color;
            this.shape = shape;
            this.sizePx = sizePx;
        }
    }

    public static final class Result {
        /** Driving field for the by-value maps below; empty for a single-symbol renderer. */
        public final String field;
        public final StrokeStyle singleStroke;
        public final FillStyle singleFill;
        public final MarkerStyle singleMarker;
        /** uniqueValue renderer: field VALUE -> style, keyed the same way AutoIconset.Result.
         * pathByValue is so a feature resolves its style on the same attribute value. */
        public final Map<String, StrokeStyle> strokeByValue;
        public final Map<String, FillStyle> fillByValue;
        public final Map<String, MarkerStyle> markerByValue;

        Result(String field, StrokeStyle singleStroke, FillStyle singleFill, MarkerStyle singleMarker,
                Map<String, StrokeStyle> strokeByValue, Map<String, FillStyle> fillByValue,
                Map<String, MarkerStyle> markerByValue) {
            this.field = field;
            this.singleStroke = singleStroke;
            this.singleFill = singleFill;
            this.singleMarker = singleMarker;
            // extract() runs on a background thread and forAutoIcons() reads these maps on the
            // main thread. Unmodifiable copies make that safe by construction rather than by
            // accident of a post() happens-before edge (Appendix A §3).
            this.strokeByValue = Collections.unmodifiableMap(
                    new LinkedHashMap<>(strokeByValue == null ? Collections.emptyMap() : strokeByValue));
            this.fillByValue = Collections.unmodifiableMap(
                    new LinkedHashMap<>(fillByValue == null ? Collections.emptyMap() : fillByValue));
            this.markerByValue = Collections.unmodifiableMap(
                    new LinkedHashMap<>(markerByValue == null ? Collections.emptyMap() : markerByValue));
        }

        public boolean isEmpty() {
            return singleStroke == null && singleFill == null && singleMarker == null
                    && strokeByValue.isEmpty() && fillByValue.isEmpty() && markerByValue.isEmpty();
        }
    }

    private AutoSymbology() {}

    /**
     * Extracts esriSMS/esriSLS/esriSFS styling from a renderer, mirroring the same
     * simple/uniqueValue/classBreaks/defaultSymbol traversal {@link AutoIconset#generate} uses.
     * Class-breaks entries can't be resolved per-feature by value (numeric-range matching isn't
     * supported here, same caveat as AutoIconset) — only a fallback symbol (defaultSymbol, or the
     * first break if there's no defaultSymbol) is extracted as a single style for that renderer
     * type. Which of stroke/fill/marker actually gets populated per entry is driven entirely by
     * each symbol's own "type" (esriSLS/esriSFS/esriSMS) rather than a geometry-type hint, so
     * mixed-symbol renderers resolve correctly without the caller needing to know the layer's
     * geometry type ahead of time.
     */
    public static Result extract(JSONObject renderer, String fieldOverride) {
        Map<String, StrokeStyle> strokeByValue = new LinkedHashMap<>();
        Map<String, FillStyle> fillByValue = new LinkedHashMap<>();
        Map<String, MarkerStyle> markerByValue = new LinkedHashMap<>();
        StrokeStyle singleStroke = null;
        FillStyle singleFill = null;
        MarkerStyle singleMarker = null;
        String field = "";

        if (renderer != null) {
            String type = renderer.optString("type", "simple");

            if ("uniqueValue".equals(type) || "uniqueValueRenderer".equals(type)) {
                field = renderer.optString("field1", renderer.optString("field", ""));
                JSONArray infos = renderer.optJSONArray("uniqueValueInfos");
                if (infos != null) {
                    for (int i = 0; i < infos.length(); i++) {
                        JSONObject info = infos.optJSONObject(i);
                        if (info == null) continue;
                        // An empty string is a legitimate uniqueValue key in ArcGIS
                        // ("unclassified"); only an *absent* value entry is skipped.
                        if (!info.has("value") || info.isNull("value")) continue;
                        String value = info.optString("value", "");
                        JSONObject symbol = info.optJSONObject("symbol");
                        if (symbol == null) continue;
                        StrokeStyle s = strokeFrom(symbol);
                        FillStyle f = fillFrom(symbol);
                        MarkerStyle m = markerFrom(symbol);
                        if (s != null) strokeByValue.put(value, s);
                        if (f != null) fillByValue.put(value, f);
                        if (m != null) markerByValue.put(value, m);
                    }
                }
                JSONObject defSymbol = renderer.optJSONObject("defaultSymbol");
                if (defSymbol != null) {
                    singleStroke = strokeFrom(defSymbol);
                    singleFill = fillFrom(defSymbol);
                    singleMarker = markerFrom(defSymbol);
                }
            } else if ("classBreaks".equals(type) || "classBreaksRenderer".equals(type)) {
                field = renderer.optString("field", "");
                // Class breaks match by numeric range, not value equality — can't be resolved
                // per-feature here (same limitation as AutoIconset); only a fallback symbol is
                // usable as a single style for the whole layer.
                JSONObject fallback = renderer.optJSONObject("defaultSymbol");
                if (fallback == null) {
                    // Deliberately NOT classBreakInfos[0]: break 0 is the lowest-value class, so
                    // using it styles the whole layer as if every feature were in the bottom
                    // bucket — a map that looks authoritative and is not (Appendix A §3).
                    // Refuse to style rather than mislead.
                    android.util.Log.d(TAG, "classBreaks renderer has no defaultSymbol — "
                            + "declining to style (numeric range matching is not implemented)");
                }
                if (fallback != null) {
                    singleStroke = strokeFrom(fallback);
                    singleFill = fillFrom(fallback);
                    singleMarker = markerFrom(fallback);
                }
            } else {
                // simple / bare symbol renderer — single style, no field
                JSONObject symbol = renderer.optJSONObject("symbol");
                singleStroke = strokeFrom(symbol);
                singleFill = fillFrom(symbol);
                singleMarker = markerFrom(symbol);
            }
        }

        if (fieldOverride != null && !fieldOverride.isEmpty()) field = fieldOverride;

        return new Result(field, singleStroke, singleFill, singleMarker,
                strokeByValue, fillByValue, markerByValue);
    }

    // -------------------------------------------------------------------------
    // Symbol JSON -> style struct
    // -------------------------------------------------------------------------

    static StrokeStyle strokeFrom(JSONObject symbol) {
        if (symbol == null) return null;
        String type = symbol.optString("type", "");
        if (!"esriSLS".equals(type)) {
            logUnhandled(type);
            return null;
        }
        // esriSLSNull is a deliberately *invisible* line. It previously fell through to
        // "solid" and a hidden boundary was drawn (Appendix A §3).
        if ("esriSLSNull".equals(symbol.optString("style", ""))) return null;
        int color = esriColor(symbol.optJSONArray("color"), Color.BLUE);
        // Esri stroke widths are in POINTS, not pixels. 1pt = 1/72in and Android's density
        // baseline is 160dpi, so 1pt ≈ 160/72 ≈ 2.222 px at density 1.0. Callers apply this as
        // a dp-equivalent width, which is the closest thing ATAK's Polyline API accepts.
        float widthDp = ptToDp((float) symbol.optDouble("width", 1.0));
        String dash = dashFromEsriStyle(symbol.optString("style", ""));
        return new StrokeStyle(color, widthDp, dash);
    }

    static FillStyle fillFrom(JSONObject symbol) {
        if (symbol == null) return null;
        String type = symbol.optString("type", "");
        if (!"esriSFS".equals(type)) {
            logUnhandled(type);
            return null;
        }
        String esriStyle = symbol.optString("style", "");
        int color = esriColor(symbol.optJSONArray("color"), Color.YELLOW);
        String style;
        if ("esriSFSNull".equals(esriStyle)) {
            style = "none";
        } else if (isHatch(esriStyle)) {
            // ATAK has no hatch fill. Rendering a hatched hazard/exclusion polygon as an opaque
            // block obscures the map underneath — on a tactical display that is a safety-relevant
            // misrepresentation, so downgrade to a low-alpha wash and say so (Appendix A §3).
            style = "solid";
            color = (color & 0x00FFFFFF) | (0x40 << 24);
            android.util.Log.d(TAG, "Hatch fill " + esriStyle
                    + " downgraded to a 25% translucent solid fill");
        } else {
            style = "solid";
        }
        StrokeStyle outline = strokeFrom(symbol.optJSONObject("outline"));
        return new FillStyle(color, style, outline);
    }

    static MarkerStyle markerFrom(JSONObject symbol) {
        if (symbol == null) return null;
        String type = symbol.optString("type", "");
        if (!"esriSMS".equals(type)) {
            logUnhandled(type);
            return null;
        }
        int color = esriColor(symbol.optJSONArray("color"), Color.RED);
        String shape = markerShapeFromEsri(symbol.optString("style", ""));
        // esriSMS.size is in points, same as esriSLS.width.
        float sizeDp = ptToDp((float) symbol.optDouble("size", 8.0));
        return new MarkerStyle(color, shape, sizeDp);
    }

    /** Points → density-independent pixels. 1pt = 1/72in; Android's dp baseline is 160dpi. */
    static float ptToDp(float pt) {
        return pt * (160f / 72f);
    }

    private static boolean isHatch(String esriStyle) {
        if (esriStyle == null) return false;
        switch (esriStyle) {
            case "esriSFSBackwardDiagonal":
            case "esriSFSCross":
            case "esriSFSDiagonalCross":
            case "esriSFSForwardDiagonal":
            case "esriSFSHorizontal":
            case "esriSFSVertical":
                return true;
            default:
                return false;
        }
    }

    /** Symbol types this extractor does not model (esriTS text, esriPFS picture fill, …). Logged
     * once per type per process so a silently-unstyled layer is at least diagnosable. */
    private static final java.util.Set<String> LOGGED_UNHANDLED =
            java.util.Collections.synchronizedSet(new java.util.HashSet<String>());

    private static void logUnhandled(String type) {
        if (type == null || type.isEmpty()) return;
        if ("esriSLS".equals(type) || "esriSFS".equals(type) || "esriSMS".equals(type)
                || "esriPMS".equals(type)) {
            return;   // esriPMS is AutoIconset's job, not ours
        }
        if (LOGGED_UNHANDLED.add(type)) {
            android.util.Log.d(TAG, "Unhandled Esri symbol type '" + type
                    + "' — no stroke/fill/marker style extracted");
        }
    }

    /** ArcGIS symbol colors are [r,g,b,a] (alpha last, 0-255); unlike DisplayConfig's own
     * esriSymbolColor helper (which drops alpha), this preserves it — translucent polygon
     * fills are common ArcGIS styling and worth carrying over. Channels are clamped: ArcGIS
     * does not guarantee them in range and an unclamped value corrupts adjacent channels. */
    static int esriColor(JSONArray arr, int fallback) {
        if (arr == null || arr.length() < 3) return fallback;
        int r = arr.optInt(0, 0);
        int g = arr.optInt(1, 0);
        int b = arr.optInt(2, 0);
        int a = arr.length() >= 4 ? arr.optInt(3, 255) : 255;
        return (clamp8(a) << 24) | (clamp8(r) << 16) | (clamp8(g) << 8) | clamp8(b);
    }

    private static int clamp8(int v) {
        return v < 0 ? 0 : (v > 255 ? 255 : v);
    }

    static String dashFromEsriStyle(String esriStyle) {
        if (esriStyle == null) return "solid";
        switch (esriStyle) {
            case "esriSLSDash":
            case "esriSLSDashDot":
            case "esriSLSDashDotDot":
                return "dash";
            case "esriSLSDot":
                return "dot";
            default:
                return "solid";
        }
    }

    private static String markerShapeFromEsri(String esriStyle) {
        if (esriStyle == null) return "circle";
        switch (esriStyle) {
            case "esriSMSSquare":   return "square";
            case "esriSMSDiamond":  return "diamond";
            case "esriSMSTriangle": return "triangle";
            case "esriSMSCross":    return "cross";
            case "esriSMSX":        return "x";
            default:                return "circle";
        }
    }
}

File: WinTAK5.6/Services/AutoSymbology.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Globalization;
using System.Linq;
using Newtonsoft.Json.Linq;

namespace FeatureLink.Services
{
    /// <summary>
    /// ArcGIS renderer → shape/marker style extractor. Line-for-line port of the ATAK plugin's
    /// <c>com.atakmap.android.featurelink.arcgis.AutoSymbology</c> (and the identical
    /// <c>CloudTAK/plugin/lib/autoSymbology.ts</c>), which had <b>no WinTAK counterpart at all</b>
    /// — audit finding C-07, the owner's field defect, and worse on this platform than any other:
    /// <c>ArcGisFeatureService.FetchLayerInfoAsync</c> read the layer metadata and discarded
    /// <c>drawingInfo</c> outright, so a layer's own published symbology was never resolved on the
    /// download path and every feature rendered with WinTAK's default marker.
    ///
    /// Extracts esriSMS (point marker shape/colour/size), esriSLS (line stroke) and esriSFS
    /// (polygon fill + outline) symbol JSON into plain style structs, keyed by renderer field
    /// value. No network I/O, no side effects — pure JSON parsing, which is why it is the first
    /// thing the new test suite covers.
    ///
    /// Semantics deliberately match the Java original exactly, including:
    ///  • class-breaks renderers contribute only a fallback style (numeric-range matching is not
    ///    resolved per feature — same documented limitation as AutoIconset);
    ///  • which of stroke/fill/marker is populated is driven by each symbol's own <c>type</c>,
    ///    not by a geometry hint, so mixed-symbol renderers resolve without the caller knowing
    ///    the layer's geometry type;
    ///  • <see cref="EsriColor"/> preserves alpha (translucent polygon fills are common ArcGIS
    ///    styling), unlike the display-config hex path which is opaque RGB.
    /// </summary>
    public static class AutoSymbology
    {
        public sealed class StrokeStyle
        {
            public Color Color { get; }
            public float WidthPx { get; }
            /// <summary>"solid" | "dash" | "dot"</summary>
            public string Dash { get; }

            public StrokeStyle(Color color, float widthPx, string dash)
            {
                Color = color; WidthPx = widthPx; Dash = dash;
            }
        }

        public sealed class FillStyle
        {
            public Color Color { get; }
            /// <summary>"solid" | "none"</summary>
            public string Style { get; }
            /// <summary>The fill symbol's own outline (esriSFS.outline is itself an esriSLS) — may be null.</summary>
            public StrokeStyle Outline { get; }

            public FillStyle(Color color, string style, StrokeStyle outline)
            {
                Color = color; Style = style; Outline = outline;
            }
        }

        public sealed class MarkerStyle
        {
            public Color Color { get; }
            /// <summary>"circle" | "square" | "diamond" | "triangle" | "cross" | "x"</summary>
            public string Shape { get; }
            public float SizePx { get; }

            public MarkerStyle(Color color, string shape, float sizePx)
            {
                Color = color; Shape = shape; SizePx = sizePx;
            }
        }

        public sealed class Result
        {
            /// <summary>Driving field for the by-value maps; empty for a single-symbol renderer.</summary>
            public string Field { get; }
            public StrokeStyle SingleStroke { get; }
            public FillStyle SingleFill { get; }
            public MarkerStyle SingleMarker { get; }
            public IDictionary<string, StrokeStyle> StrokeByValue { get; }
            public IDictionary<string, FillStyle> FillByValue { get; }
            public IDictionary<string, MarkerStyle> MarkerByValue { get; }

            public Result(string field, StrokeStyle singleStroke, FillStyle singleFill, MarkerStyle singleMarker,
                IDictionary<string, StrokeStyle> strokeByValue, IDictionary<string, FillStyle> fillByValue,
                IDictionary<string, MarkerStyle> markerByValue)
            {
                Field = field ?? string.Empty;
                SingleStroke = singleStroke;
                SingleFill = singleFill;
                SingleMarker = singleMarker;
                StrokeByValue = strokeByValue ?? new Dictionary<string, StrokeStyle>();
                FillByValue = fillByValue ?? new Dictionary<string, FillStyle>();
                MarkerByValue = markerByValue ?? new Dictionary<string, MarkerStyle>();
            }

            public bool IsEmpty =>
                SingleStroke == null && SingleFill == null && SingleMarker == null
                && StrokeByValue.Count == 0 && FillByValue.Count == 0 && MarkerByValue.Count == 0;

            /// <summary>True when the renderer carries polyline/polygon (esriSLS/esriSFS) styling —
            /// the part of a Portal-authored config that was previously dropped on WinTAK with no
            /// warning and no log line (Appendix F §5's concrete interop break).</summary>
            public bool HasShapeStyling =>
                SingleStroke != null || SingleFill != null
                || StrokeByValue.Count > 0 || FillByValue.Count > 0;
        }

        /// <summary>One classified renderer category: the attribute value it matches, its label,
        /// and its symbol.</summary>
        public sealed class ValueEntry
        {
            public string Value { get; set; }
            public string Label { get; set; }
            public JObject Symbol { get; set; }
        }

        /// <summary>
        /// Yields a renderer's classified categories, in renderer order, from EITHER unique-value
        /// layout.
        ///
        /// <para><b>Why both.</b> ArcGIS has two shapes for the same thing. The classic one is a
        /// flat <c>uniqueValueInfos</c> array. Modern ArcGIS Online publishes
        /// <c>uniqueValueGroups[].classes[]</c> instead, where each class carries
        /// <c>values: [["A"]]</c> — an array of arrays, because a unique-value renderer can key on
        /// up to three fields.</para>
        ///
        /// <para>Reading only the classic shape is silently catastrophic rather than merely
        /// incomplete: a modern renderer's categories are simply not seen, so the extractors fall
        /// through to <c>defaultSymbol</c> and every feature gets one symbol. That is exactly what
        /// happened in the field — a layer showing twelve distinct symbols in ArcGIS produced a
        /// single orange dot on every marker, with no error anywhere, because the one thing we
        /// did find was the default.</para>
        ///
        /// <para>Order is preserved: <c>AUTO-ICONSET-SPEC.md</c> §5.3 assigns filename collision
        /// suffixes in renderer array order, and every platform must agree on it.</para>
        /// </summary>
        public static List<ValueEntry> EnumerateValueEntries(JObject renderer)
        {
            var entries = new List<ValueEntry>();
            if (renderer == null) return entries;

            // Classic: flat uniqueValueInfos.
            if (renderer["uniqueValueInfos"] is JArray infos)
            {
                foreach (var info in infos.OfType<JObject>())
                {
                    var symbol = info["symbol"] as JObject;
                    if (symbol == null) continue;
                    entries.Add(new ValueEntry
                    {
                        Value = OptString(info, "value", null),
                        Label = OptString(info, "label", null),
                        Symbol = symbol,
                    });
                }
            }

            // Modern: uniqueValueGroups -> classes -> values[][].
            if (renderer["uniqueValueGroups"] is JArray groups)
            {
                // Multi-field renderers join their key parts with this; ArcGIS defaults to ",".
                string delimiter = OptString(renderer, "fieldDelimiter", ",");

                foreach (var group in groups.OfType<JObject>())
                {
                    if (!(group["classes"] is JArray classes)) continue;
                    foreach (var cls in classes.OfType<JObject>())
                    {
                        var symbol = cls["symbol"] as JObject;
                        if (symbol == null) continue;
                        string label = OptString(cls, "label", null);

                        // One class can cover SEVERAL values sharing a symbol; each becomes its
                        // own entry so per-value resolution finds any of them.
                        var values = cls["values"] as JArray;
                        if (values == null || values.Count == 0)
                        {
                            entries.Add(new ValueEntry { Value = null, Label = label, Symbol = symbol });
                            continue;
                        }

                        foreach (var v in values)
                        {
                            string composed;
                            if (v is JArray parts)
                                composed = string.Join(delimiter,
                                    parts.Select(p => p == null || p.Type == JTokenType.Null
                                        ? string.Empty
                                        : p.ToString(Newtonsoft.Json.Formatting.None).Trim('"')));
                            else
                                composed = v.ToString(Newtonsoft.Json.Formatting.None).Trim('"');

                            entries.Add(new ValueEntry
                            {
                                Value = composed,
                                Label = label,
                                Symbol = symbol,
                            });
                        }
                    }
                }
            }

            return entries;
        }

        /// <summary>
        /// Extracts esriSMS/esriSLS/esriSFS styling from a renderer, mirroring the
        /// simple/uniqueValue/classBreaks/defaultSymbol traversal AutoIconset uses.
        /// </summary>
        /// <param name="renderer">the <c>drawingInfo.renderer</c> object, or a Web-Map per-layer
        /// style override the caller already resolved (AUTO-ICONSET-SPEC.md §2.3).</param>
        /// <param name="fieldOverride">renderer field override; wins over the renderer's own field.</param>
        public static Result Extract(JObject renderer, string fieldOverride = null)
        {
            var strokeByValue = new Dictionary<string, StrokeStyle>(StringComparer.Ordinal);
            var fillByValue = new Dictionary<string, FillStyle>(StringComparer.Ordinal);
            var markerByValue = new Dictionary<string, MarkerStyle>(StringComparer.Ordinal);
            StrokeStyle singleStroke = null;
            FillStyle singleFill = null;
            MarkerStyle singleMarker = null;
            string field = string.Empty;

            if (renderer != null)
            {
                string type = OptString(renderer, "type", "simple");

                if (type == "uniqueValue" || type == "uniqueValueRenderer")
                {
                    field = OptString(renderer, "field1", null) ?? OptString(renderer, "field", string.Empty);

                    // Covers BOTH uniqueValueInfos and uniqueValueGroups/classes — see
                    // EnumerateValueEntries. Reading only the former made a modern ArcGIS Online
                    // renderer look like it had no categories at all.
                    foreach (var entry in EnumerateValueEntries(renderer))
                    {
                        string value = entry.Value;
                        if (string.IsNullOrEmpty(value) || entry.Symbol == null) continue;
                        var s = StrokeFrom(entry.Symbol);
                        var f = FillFrom(entry.Symbol);
                        var m = MarkerFrom(entry.Symbol);
                        if (s != null) strokeByValue[value] = s;
                        if (f != null) fillByValue[value] = f;
                        if (m != null) markerByValue[value] = m;
                    }
                    if (renderer["defaultSymbol"] is JObject defSymbol)
                    {
                        singleStroke = StrokeFrom(defSymbol);
                        singleFill = FillFrom(defSymbol);
                        singleMarker = MarkerFrom(defSymbol);
                    }
                }
                else if (type == "classBreaks" || type == "classBreaksRenderer")
                {
                    field = OptString(renderer, "field", string.Empty);
                    // Class breaks match by numeric range, not value equality — not resolvable per
                    // feature here (same limitation as AutoIconset); only a fallback symbol is
                    // usable as a single style for the whole layer.
                    var fallback = renderer["defaultSymbol"] as JObject;
                    if (fallback == null)
                    {
                        // Deliberately NOT classBreakInfos[0], which this used to fall back to.
                        // Break 0 is the LOWEST-value class, so using it styles the whole layer as
                        // if every feature sat in the bottom bucket — a map that looks
                        // authoritative and is not. ATAK refuses here for exactly this reason;
                        // WinTAK doing otherwise was both a divergence and the more dangerous of
                        // the two behaviours. Refuse to style rather than mislead.
                        Log.Info("classBreaks renderer has no defaultSymbol — declining to style "
                                 + "(numeric range matching is not implemented).");
                    }
                    if (fallback != null)
                    {
                        singleStroke = StrokeFrom(fallback);
                        singleFill = FillFrom(fallback);
                        singleMarker = MarkerFrom(fallback);
                    }
                }
                else
                {
                    // simple / bare symbol renderer — single style, no field
                    var symbol = renderer["symbol"] as JObject;
                    singleStroke = StrokeFrom(symbol);
                    singleFill = FillFrom(symbol);
                    singleMarker = MarkerFrom(symbol);
                }
            }

            if (!string.IsNullOrEmpty(fieldOverride)) field = fieldOverride;

            return new Result(field, singleStroke, singleFill, singleMarker,
                strokeByValue, fillByValue, markerByValue);
        }

        // -------------------------------------------------------------------------
        // Symbol JSON → style struct
        // -------------------------------------------------------------------------

        internal static StrokeStyle StrokeFrom(JObject symbol)
        {
            if (symbol == null || OptString(symbol, "type", string.Empty) != "esriSLS") return null;
            Color color = EsriColor(symbol["color"] as JArray, Color.Blue);
            float width = (float)OptDouble(symbol, "width", 1.0);
            string dash = DashFromEsriStyle(OptString(symbol, "style", string.Empty));
            return new StrokeStyle(color, width, dash);
        }

        internal static FillStyle FillFrom(JObject symbol)
        {
            if (symbol == null || OptString(symbol, "type", string.Empty) != "esriSFS") return null;
            Color color = EsriColor(symbol["color"] as JArray, Color.Yellow);
            string style = OptString(symbol, "style", string.Empty) == "esriSFSNull" ? "none" : "solid";
            StrokeStyle outline = StrokeFrom(symbol["outline"] as JObject);
            return new FillStyle(color, style, outline);
        }

        internal static MarkerStyle MarkerFrom(JObject symbol)
        {
            if (symbol == null || OptString(symbol, "type", string.Empty) != "esriSMS") return null;
            Color color = EsriColor(symbol["color"] as JArray, Color.Red);
            string shape = MarkerShapeFromEsri(OptString(symbol, "style", string.Empty));
            float size = (float)OptDouble(symbol, "size", 8.0);
            return new MarkerStyle(color, shape, size);
        }

        /// <summary>ArcGIS symbol colours are <c>[r,g,b,a]</c> (alpha last, 0-255). Alpha is
        /// preserved here — translucent polygon fills are common ArcGIS styling.</summary>
        internal static Color EsriColor(JArray arr, Color fallback)
        {
            if (arr == null || arr.Count < 3) return fallback;
            int r = OptInt(arr, 0, 0);
            int g = OptInt(arr, 1, 0);
            int b = OptInt(arr, 2, 0);
            int a = arr.Count >= 4 ? OptInt(arr, 3, 255) : 255;
            return Color.FromArgb(Clamp255(a), Clamp255(r), Clamp255(g), Clamp255(b));
        }

        internal static string DashFromEsriStyle(string esriStyle)
        {
            switch (esriStyle)
            {
                case "esriSLSDash":
                case "esriSLSDashDot":
                case "esriSLSDashDotDot":
                    return "dash";
                case "esriSLSDot":
                    return "dot";
                default:
                    return "solid";
            }
        }

        internal static string MarkerShapeFromEsri(string esriStyle)
        {
            switch (esriStyle)
            {
                case "esriSMSSquare": return "square";
                case "esriSMSDiamond": return "diamond";
                case "esriSMSTriangle": return "triangle";
                case "esriSMSCross": return "cross";
                case "esriSMSX": return "x";
                default: return "circle";
            }
        }

        // -------------------------------------------------------------------------
        // Culture-invariant JSON accessors
        // -------------------------------------------------------------------------
        // Every numeric read below pins InvariantCulture. The audit measured zero
        // CultureInfo.InvariantCulture uses in either tree; a de-DE machine round-tripping a
        // JSON numeric through the default culture is the classic silent-wrong-symbology bug.

        private static string OptString(JObject o, string key, string fallback)
        {
            var t = o?[key];
            if (t == null || t.Type == JTokenType.Null) return fallback;
            return t.Type == JTokenType.String ? (string)t : t.ToString(Newtonsoft.Json.Formatting.None).Trim('"');
        }

        private static double OptDouble(JObject o, string key, double fallback)
        {
            var t = o?[key];
            if (t == null || t.Type == JTokenType.Null) return fallback;
            if (t.Type == JTokenType.Integer || t.Type == JTokenType.Float) return (double)t;
            double parsed;
            return double.TryParse((string)t, NumberStyles.Float, CultureInfo.InvariantCulture, out parsed)
                ? parsed : fallback;
        }

        private static int OptInt(JArray arr, int index, int fallback)
        {
            if (arr == null || index >= arr.Count) return fallback;
            var t = arr[index];
            if (t == null || t.Type == JTokenType.Null) return fallback;
            if (t.Type == JTokenType.Integer || t.Type == JTokenType.Float) return (int)Math.Round((double)t);
            int parsed;
            return int.TryParse((string)t, NumberStyles.Integer, CultureInfo.InvariantCulture, out parsed)
                ? parsed : fallback;
        }

        private static int Clamp255(int v) => v < 0 ? 0 : (v > 255 ? 255 : v);
    }
}
